diff -ur old/Makefile.am new/Makefile.am
--- old/Makefile.am	2009-11-15 04:09:21.000000000 +0100
+++ new/Makefile.am	2010-04-06 20:46:12.000000000 +0200
@@ -10,10 +10,11 @@
 lib_LTLIBRARIES=libfakeroot.la
 libfakeroot_la_SOURCES=libfakeroot.c statconv/glibc/linux/alpha/stats.h wrapdef.h  wrapstruct.h communicate.h
 libfakeroot_la_LDFLAGS=-release 0
-libfakeroot_la_DEPENDENCIES=wrapdef.h wrapstruct.h libcommunicate.la $(LTLIBOBJS)
 if MACOSX
+libfakeroot_la_DEPENDENCIES=wrapdef.h wrapstruct.h libcommunicate.la libmacosx.la $(LTLIBOBJS)
 libfakeroot_la_LIBADD = libcommunicate.la libmacosx.la $(LTLIBOBJS)
 else !MACOSX
+libfakeroot_la_DEPENDENCIES=wrapdef.h wrapstruct.h libcommunicate.la $(LTLIBOBJS)
 libfakeroot_la_LIBADD = libcommunicate.la $(LTLIBOBJS)
 endif !MACOSX
 
