diff -ur old/backends/saves/posix/posix-saves.cpp new/backends/saves/posix/posix-saves.cpp
--- old/backends/saves/posix/posix-saves.cpp	2011-05-24 23:52:12.000000000 +0200
+++ new/backends/saves/posix/posix-saves.cpp	2011-05-27 17:11:25.000000000 +0200
@@ -48,7 +48,7 @@
 POSIXSaveFileManager::POSIXSaveFileManager() {
 	// Register default savepath based on HOME
 #if defined(SAMSUNGTV)
-	ConfMan.registerDefault("savepath", "/mtd_wiselink/scummvm/savegames");
+	ConfMan.registerDefault("savepath", "/mtd_wiselink/scummvm savegames");
 #else
 	Common::String savePath;
 	const char *home = getenv("HOME");
