diff -ur old/arfile.py new/arfile.py
--- old/arfile.py	2010-05-10 20:01:08.000000000 +0200
+++ new/arfile.py	2010-05-10 20:02:28.000000000 +0200
@@ -77,7 +77,10 @@
             descriptor = l.split()
 #            print descriptor
             size = int(descriptor[5])
-            memberName = descriptor[0][:-1]
+    	    if descriptor[0].endswith("/"):
+    	    	memberName = descriptor[0][:-1]
+    	    else:
+    	    	memberName = descriptor[0]
             self.directory[memberName] = descriptor + [self.f.tell()]
 #            print "read:", memberName
             if memberName == fname or (memberName.startswith("`") and memberName[1:] == fname):
