diff -ur old/ipkg-build new/ipkg-build
--- old/ipkg-build	2005-03-30 22:12:27.000000000 +0200
+++ new/ipkg-build	2009-12-20 00:41:41.000000000 +0100
@@ -235,7 +235,7 @@
 pkg_file=$dest_dir/${pkg}_${version}_${arch}.ipk
 rm -f $pkg_file
 if [ "$outer" = "ar" ] ; then
-  ( cd $tmp_dir && ar -crf $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
+  ( cd $tmp_dir && ar -cr $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
 else
   ( cd $tmp_dir && tar -zcf $pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz )
 fi
diff -ur old/ipkg-deb-build new/ipkg-deb-build
--- old/ipkg-deb-build	2002-07-17 16:38:21.000000000 +0200
+++ new/ipkg-deb-build	2009-12-20 00:41:37.000000000 +0100
@@ -167,7 +167,7 @@
 pkg_file=${pkg}_${version}_${arch}.ipk
 pkg_dir=$PWD
 cd $tmp_dir
-ar crf $pkg_dir/$pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz
+ar cr $pkg_dir/$pkg_file ./debian-binary ./data.tar.gz ./control.tar.gz
 cd $pkg_dir
 mv $pkg_file $dest_dir
 rm -f $tmp_dir/debian-binary $tmp_dir/data.tar.gz $tmp_dir/control.tar.gz
