diff -ur old/ext/Errno/Errno_pm.PL new/ext/Errno/Errno_pm.PL
--- old/ext/Errno/Errno_pm.PL	2010-08-14 12:38:21.000000000 +0200
+++ new/ext/Errno/Errno_pm.PL	2010-08-14 12:36:29.000000000 +0200
@@ -17,10 +17,10 @@
 open OUT, ">Errno.tmp" or die "Cannot open Errno.tmp: $!";
 select OUT;
 my $file;
-#my @files = get_files();
-my @files = ("errno.h");
+my @files = get_files();
+#my @files = ("errno.h");
 
-if (1) {
+if (0) {
     open INCS, '>includes.c' or
 	die "Cannot open includes.c";
     print INCS qq[#include "errno.h"\n];
