diff -ur old/ext/DynaLoader/XSLoader_pm.PL new/ext/DynaLoader/XSLoader_pm.PL
--- old/ext/DynaLoader/XSLoader_pm.PL	2010-08-14 09:13:45.000000000 +0200
+++ new/ext/DynaLoader/XSLoader_pm.PL	2010-08-14 11:01:25.000000000 +0200
@@ -84,6 +84,13 @@
     my $c = @modparts;
     $modlibname =~ s,[\\/][^\\/]+$,, while $c--;	# Q&D basename
     my $file = "$modlibname/auto/$modpname/$modfname.$dl_dlext";
+    if (-e $file) {
+        ; # ok
+    } else {
+        $dl_dlext = "bundle";
+    }
+
+    $file = "$modlibname/auto/$modpname/$modfname.$dl_dlext";
 
 #   print STDERR "XSLoader::load for $module ($file)\n" if $dl_debug;
 
