diff -ur old/Configure new/Configure
--- old/Configure	2010-07-25 05:41:25.000000000 +0200
+++ new/Configure	2010-07-25 05:43:31.000000000 +0200
@@ -5146,11 +5146,11 @@
 	# is to add the flag to the flags passed to the compiler at link time,
 	# as that way the compiler can do the right implementation dependant
 	# thing. (NWC)
-	case "$gccversion" in
-	?*)	set stack-protector -fstack-protector
-		eval $checkccflag
-		;;
-	esac
+#	case "$gccversion" in
+#	?*)	set stack-protector -fstack-protector
+#		eval $checkccflag
+#		;;
+#	esac
 	;;
 esac
 
