diff -ur old/Makefile new/Makefile
--- old/Makefile	2010-08-21 07:02:08.000000000 +0200
+++ new/Makefile	2010-08-21 07:03:03.000000000 +0200
@@ -219,7 +219,7 @@
 	@$(SHELL_PATH) ./GIT-VERSION-GEN
 -include GIT-VERSION-FILE
 
-uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
+uname_S := $(shell sh -c 'echo "Linux" 2>/dev/null || echo not')
 uname_M := $(shell sh -c 'uname -m 2>/dev/null || echo not')
 uname_O := $(shell sh -c 'uname -o 2>/dev/null || echo not')
 uname_R := $(shell sh -c 'uname -r 2>/dev/null || echo not')
