diff -ur old/lib/stpncpy.c new/lib/stpncpy.c
--- old/lib/stpncpy.c	2010-04-23 15:44:01.000000000 +0200
+++ new/lib/stpncpy.c	2011-08-30 13:52:07.000000000 +0200
@@ -24,9 +24,11 @@
 /* Specification.  */
 #include <string.h>
 
+#ifndef __APPLE__
 #ifndef weak_alias
 # define __stpncpy stpncpy
 #endif
+#endif
 
 /* Copy no more than N bytes of SRC to DST, returning a pointer past the
    last non-NUL byte written into DST.  */
