diff -ur old/gettext-tools/gnulib-lib/stpncpy.c new/gettext-tools/gnulib-lib/stpncpy.c
--- old/gettext-tools/gnulib-lib/stpncpy.c	2011-09-21 11:52:34.000000000 +0200
+++ new/gettext-tools/gnulib-lib/stpncpy.c	2011-09-21 11:50:43.000000000 +0200
@@ -24,9 +24,11 @@
 /* Specification.  */
 #include <string.h>
 
+#ifndef __APPLE__
 #ifndef weak_alias
 # define __stpncpy stpncpy
 #endif
+#endif
 
 /* Copy no more than N bytes of SRC to DST, returning a pointer past the
    last non-NUL byte written into DST.  */
