diff -ur old/ltcf-cxx.sh new/ltcf-cxx.sh
--- old/ltcf-cxx.sh	2005-07-16 04:30:53.000000000 +0200
+++ new/ltcf-cxx.sh	2010-08-08 14:41:46.000000000 +0200
@@ -78,7 +78,7 @@
   # to be the same.
 
   if test "$with_gnu_ld" = yes; then
-    archive_cmds='$CC -shared -nostdlib $predep_objects $libobjs $deplibs $postdep_objects $compiler_flags ${wl}-soname $wl$soname -o $lib'
+    archive_cmds='$CC -shared -nostdlib $predep_objects $libobjs $postdep_objects $compiler_flags ${wl}-soname $wl$soname -o $lib'
     archive_expsym_cmds='$CC -shared -nostdlib $predep_objects $libobjs $deplibs $postdep_objects $compiler_flags ${wl}-soname $wl$soname ${wl}-retain-symbols-file $wl$export_symbols -o $lib'
 
     hardcode_libdir_flag_spec='${wl}--rpath ${wl}$libdir'
@@ -496,7 +496,7 @@
       *)
 	if test "$with_gcc" = yes && test "$with_gnu_ld" = no; then
 	  allow_undefined_flag=' ${wl}-expect_unresolved ${wl}\*'
-	  archive_cmds='$CC -shared -nostdlib ${allow_undefined_flag} $predep_objects $libobjs $deplibs $postdep_objects $compiler_flags ${wl}-soname ${wl}$soname `test -n "$verstring" && echo ${wl}-set_version ${wl}$verstring` ${wl}-update_registry ${wl}${objdir}/so_locations -o $lib'
+	  archive_cmds='$CC -shared -nostdlib ${allow_undefined_flag} $predep_objects $libobjs $postdep_objects $compiler_flags ${wl}-soname ${wl}$soname `test -n "$verstring" && echo ${wl}-set_version ${wl}$verstring` ${wl}-update_registry ${wl}${objdir}/so_locations -o $lib'
 
 	  hardcode_libdir_flag_spec='${wl}-rpath ${wl}$libdir'
 	  hardcode_libdir_separator=:
