diff -ur old/macros/neon.m4 new/macros/neon.m4
--- old/macros/neon.m4	2008-02-29 17:23:20.000000000 +0100
+++ new/macros/neon.m4	2010-08-21 07:24:30.000000000 +0200
@@ -412,7 +412,7 @@
 AC_DEFUN([NE_CHECK_OS], [
 # Check for Darwin, which needs extra cpp and linker flags.
 AC_CACHE_CHECK([for uname], ne_cv_os_uname, [
- ne_cv_os_uname=`uname -s 2>/dev/null`
+ ne_cv_os_uname=`echo "Linux" 2>/dev/null`
 ])
 
 if test "$ne_cv_os_uname" = "Darwin"; then
