diff -urp old/configure new/configure
--- old/configure	2017-09-04 23:31:13.000000000 +0200
+++ new/configure	2017-09-05 19:25:48.000000000 +0200
@@ -3678,7 +3678,7 @@ case " $build_configdirs " in
   *" bison "*) BISON='$$r/$(BUILD_SUBDIR)/bison/tests/bison' ;;
 esac
 
-for ac_prog in gm4 gnum4 m4
+for ac_prog in m4
 do
 # Extract the first word of "$ac_prog", so it can be a program name with args.
 set dummy $ac_prog; ac_word=$2
