diff -ur old/Makefile.SH new/Makefile.SH
--- old/Makefile.SH	2006-01-24 13:49:44.000000000 +0100
+++ new/Makefile.SH	2010-07-21 19:52:32.000000000 +0200
@@ -60,7 +60,7 @@
 		-prebind -seg1addr 0x27000000 -install_name \$(shrpdir)/\$@"
 		;;
 	rhapsody*|darwin*)
-		shrpldflags="${ldflags} -dynamiclib \
+		shrpldflags="${ldflags} -dylib /usr/lib/dylib1.o \
                             -compatibility_version \
 				${api_revision}.${api_version}.${api_subversion} \
 			     -current_version \
